docs/concepts/filters.md

```diff
@@ -0,0 +1,257 @@
+---
+title: Filters
+author: halter73
+description: MCP Server Handler Filters
+uid: filters
+---
+
+# MCP Server Handler Filters
+
+For each handler type in the MCP Server, there are corresponding `AddXXXFilter` methods in `McpServerBuilderExtensions.cs` that allow you to add filters to the handler pipeline. The filters are stored in `McpServerOptions.Filters` and applied during server configuration.
+
+## Available Filter Methods
+
+The following filter methods are available:
+
+- `AddListResourceTemplatesFilter` - Filter for list resource templates handlers
+- `AddListToolsFilter` - Filter for list tools handlers
+- `AddCallToolFilter` - Filter for call tool handlers
+- `AddListPromptsFilter` - Filter for list prompts handlers
+- `AddGetPromptFilter` - Filter for get prompt handlers
+- `AddListResourcesFilter` - Filter for list resources handlers
+- `AddReadResourceFilter` - Filter for read resource handlers
+- `AddCompleteFilter` - Filter for completion handlers
+- `AddSubscribeToResourcesFilter` - Filter for resource subscription handlers
+- `AddUnsubscribeFromResourcesFilter` - Filter for resource unsubscription handlers
+- `AddSetLoggingLevelFilter` - Filter for logging level handlers
+
+## Usage
+
+Filters are functions that take a handler and return a new handler, allowing you to wrap the original handler with additional functionality:
+
+```csharp
+services.AddMcpServer()
+    .WithListToolsHandler(async (context, cancellationToken) =>
+    {
+        // Your base handler logic
+        return new ListToolsResult { Tools = GetTools() };
+    })
+    .AddListToolsFilter(next => async (context, cancellationToken) =>
```

Comment:
I'm confused with this signature. Why is the delegate returning another delegate, as opposed to this being, e.g.
```csharp
(async (next, context, cancellationToken) =>
```
or just
```csharp
(async (context, cancellationToken) =>
```
with `next` part of `context`, e.g. `context.Next`?

Reply:
We could put the next function inside the context. It would require us to create a new context type rather than reusing `RequestContext<TParams>` directly though. A delegate returning another delegate is also the pattern we use in in [ASP.NET Core request middleware](https://github.com/dotnet/aspnetcore/blob/8d1e32deea24273caebf1349ca55884ff26f03e1/src/Http/Http/src/Builder/ApplicationBuilder.cs#L23), [Endpoint filters](https://github.com/dotnet/aspnetcore/blob/8d1e32deea24273caebf1349ca55884ff26f03e1/src/Http/Http.Abstractions/src/Extensions/EndpointBuilder.cs#L13), and [Kestrel's connection middleware](https://github.com/dotnet/aspnetcore/blob/8d1e32deea24273caebf1349ca55884ff26f03e1/src/Servers/Kestrel/Core/src/ListenOptions.cs#L21), so it's pretty standard.

Reply:
We can also add an extension method that add the `(async (next, context, cancellationToken) =>` version. We do that for ASP.NET Core request middleware in [UseExtensions](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.builder.useextensions?view=aspnetcore-9.0).